### ISSUE TITLE
Potential memory leak using `Table` and rocksdb

### ISSUE THREAD

**Issue by ipeluffo** [robinhood/faust#383]
## Checklist

- [x] I have included information about relevant versions
- [ ] I have verified that the issue persists when using the `master` branch of Faust.

## Steps to reproduce

The source code to reproduce the issue can be found in: https://github.com/ipeluffo/faust-table-test .

On one terminal run:
```bash
pipenv run faust -A app worker -l info
```

On another terminal run:
```bash
pipenv run faust -A commands send-events --max-events 1500000
```

## Expected behavior

Agent should process all events and store data in a faust table without increasing the memory usage too much.

## Actual behavior

Memory usage increase indefinitely.

## Additional Information

Before running the worker, the objects below were the most 20 most common (info taken using `objgraph`):
```
objgraph.show_most_common_types(limit=20)
[2019-07-11 22:08:00,847: WARNING]: function                   19994
[2019-07-11 22:08:00,848: WARNING]: dict                       15772
[2019-07-11 22:08:00,848: WARNING]: tuple                      12316
[2019-07-11 22:08:00,848: WARNING]: list                       4986
[2019-07-11 22:08:00,848: WARNING]: weakref                    4952
[2019-07-11 22:08:00,848: WARNING]: type                       2775
[2019-07-11 22:08:00,849: WARNING]: getset_descriptor          2734
[2019-07-11 22:08:00,849: WARNING]: cell                       2074
[2019-07-11 22:08:00,849: WARNING]: builtin_function_or_method 1673
[2019-07-11 22:08:00,849: WARNING]: property                   1529
[2019-07-11 22:08:00,849: WARNING]: wrapper_descriptor         1512
[2019-07-11 22:08:00,849: WARNING]: method_descriptor          1438
[2019-07-11 22:08:00,850: WARNING]: module                     1136
[2019-07-11 22:08:00,850: WARNING]: ModuleSpec                 1132
[2019-07-11 22:08:00,850: WARNING]: set                        1074
[2019-07-11 22:08:00,850: WARNING]: _GenericAlias              1055
[2019-07-11 22:08:00,850: WARNING]: SourceFileLoader           1039
[2019-07-11 22:08:00,850: WARNING]: member_descriptor          804
[2019-07-11 22:08:00,851: WARNING]: frozenset                  709
[2019-07-11 22:08:00,851: WARNING]: ABCMeta                    559
```

After processing 1.5 million events:
```
objgraph.show_most_common_types(limit=20)
[2019-07-11 22:35:27,255: WARNING]: dict                       1515758
[2019-07-11 22:35:27,255: WARNING]: method                     1500518
[2019-07-11 22:35:27,255: WARNING]: FutureMessage              1500000
[2019-07-11 22:35:27,255: WARNING]: PendingMessage             1500000
[2019-07-11 22:35:27,256: WARNING]: function                   20000
[2019-07-11 22:35:27,256: WARNING]: tuple                      11588
[2019-07-11 22:35:27,256: WARNING]: weakref                    5173
[2019-07-11 22:35:27,256: WARNING]: list                       5117
[2019-07-11 22:35:27,256: WARNING]: type                       2775
[2019-07-11 22:35:27,256: WARNING]: getset_descriptor          2734
[2019-07-11 22:35:27,256: WARNING]: cell                       2082
[2019-07-11 22:35:27,256: WARNING]: builtin_function_or_method 1891
[2019-07-11 22:35:27,256: WARNING]: property                   1529
[2019-07-11 22:35:27,257: WARNING]: wrapper_descriptor         1516
[2019-07-11 22:35:27,257: WARNING]: method_descriptor          1441
[2019-07-11 22:35:27,257: WARNING]: module                     1136
[2019-07-11 22:35:27,258: WARNING]: ModuleSpec                 1132
[2019-07-11 22:35:27,258: WARNING]: set                        1083
[2019-07-11 22:35:27,258: WARNING]: _GenericAlias              1055
[2019-07-11 22:35:27,258: WARNING]: SourceFileLoader           1039
```

Memory usage before processing events:
<img width="665" alt="Faust 01 - Before processing events" src="https://user-images.githubusercontent.com/889705/61088838-36211000-a431-11e9-853e-b7a96a167e36.png">

Memory usage after processing 1.5 million events:
<img width="669" alt="Faust 02 - After processing events" src="https://user-images.githubusercontent.com/889705/61088837-36211000-a431-11e9-8a3f-0a57cc4068ce.png">

Memory usage after restarting the worker:
<img width="666" alt="Faust 03 - After restarting worker" src="https://user-images.githubusercontent.com/889705/61088835-35887980-a431-11e9-922c-26e4d64ee266.png">

# Versions

* Python version: 3.7.3
* Faust version: 1.7.1
* Operating system: macOS Mojave 10.14.5
* Kafka version: 2.2.0
* RocksDB version (if applicable)

**Comment by StephenSorriaux** [comment]
Hi,

I encountered similar issue and I was able to reduce the memory usage by changing the settings of the underlying Rocksdb database. By default, cache size is set to 2GB (https://github.com/robinhood/faust/blob/master/faust/stores/rocksdb.py#L43) which can explain what you are saying.
Since the 1.7.1 version, you should be able to pass an `options` object to the table and change those values (see https://github.com/robinhood/faust/commit/0085eeaccb83bdb67c3ca17e91d1c0a390288963). Also, if your table keys are somehow "big", you can change the value of the associated cache (see https://github.com/robinhood/faust/issues/372 and https://github.com/robinhood/faust/issues/372)

**Comment by ipeluffo** [comment]
@StephenSorriaux , thanks for your answer. I'll repeat the experiment changing the cache size (`block_cache_size` option) and see if faust stops using memory beyond the limit.

**Comment by ipeluffo** [comment]
I repeated the test only changing the value of `block_cache_size`, changing the table creation as:
```python
events_table = app.Table(name='events_table', default=int, options=dict(
    block_cache_size=256 ** 3,
))
```

Unfortunately, I got the same results:
```
objgraph.show_most_common_types(limit=20)
[2019-07-12 17:20:34,675: WARNING]: dict                       1515774
[2019-07-12 17:20:34,675: WARNING]: method                     1500519
[2019-07-12 17:20:34,675: WARNING]: FutureMessage              1500000
[2019-07-12 17:20:34,675: WARNING]: PendingMessage             1500000
[2019-07-12 17:20:34,675: WARNING]: function                   20021
[2019-07-12 17:20:34,675: WARNING]: tuple                      11608
[2019-07-12 17:20:34,676: WARNING]: weakref                    5175
[2019-07-12 17:20:34,676: WARNING]: list                       5123
[2019-07-12 17:20:34,676: WARNING]: type                       2775
[2019-07-12 17:20:34,676: WARNING]: getset_descriptor          2734
[2019-07-12 17:20:34,676: WARNING]: cell                       2082
[2019-07-12 17:20:34,676: WARNING]: builtin_function_or_method 1892
[2019-07-12 17:20:34,676: WARNING]: property                   1529
[2019-07-12 17:20:34,676: WARNING]: wrapper_descriptor         1516
[2019-07-12 17:20:34,676: WARNING]: method_descriptor          1441
[2019-07-12 17:20:34,676: WARNING]: module                     1136
[2019-07-12 17:20:34,676: WARNING]: ModuleSpec                 1132
[2019-07-12 17:20:34,676: WARNING]: set                        1084
[2019-07-12 17:20:34,676: WARNING]: _GenericAlias              1055
[2019-07-12 17:20:34,677: WARNING]: SourceFileLoader           1039
```

<img width="615" alt="Screenshot 2019-07-12 at 17 22 19 (2)" src="https://user-images.githubusercontent.com/889705/61143310-bc893080-a4c9-11e9-9593-e4d25136b4fa.png">

**Comment by ask** [comment]
I wonder why FutureMessage/PendingMessage is at the top of the list?

When you do `await topic.send()` this will create FutureMessage objects, but they should be garbage collected once they are sent.

It's possible this is `app.send_soon()` leaking, and these are changelog messages being produced.

**Comment by ask** [comment]
It's app.send_soon! It's not draining the buffer fast enough, and when stream processing is fast it will build up and cause trouble.

**Comment by ipeluffo** [comment]
@ask , I'm a bit confused because you're saying that the issue is in the send while the objects and high memory usage is in the worker/agent receiving the events and storing them in the table.

**Comment by ask** [comment]
No, it was back pressure in the producer `send_soon` buffer.

Modifying the table calls ``__getitem__`` which is non-asynchronous, so to produce a message into the changelog we have to call `topic.send_soon()`.

This adds to a buffer, and a background task is responsible for draining this buffer and actually producing the messages.

The problem is that the stream processing is going too fast, so the buffer continues to grow
and grow.

The commit above fixes it (4c541cb): we have stream processing wait for the buffer
to be of an acceptable size before processing any more messages.

**Comment by yungchin** [comment]
@ask I ran into much the same problem and found this issue after I discovered 4c541cb fixed the problem. The thing is though, the test script I created doesn't create much stream activity at all, so I don't understand how the fix works - can I ask you about that here?

My test script just updates a table every second:
```
from datetime import datetime

import faust


app = faust.App("debug-changelog")
topic = app.topic("topic", internal=True)
table = app.Table("table")


@app.timer(interval=1.)
async def every_second():
    await topic.send(value=str(datetime.utcnow()))


@app.agent(topic)
async def write_table(stream):
    async for data in stream:
        table["now"] = data


app.main()
```
and then I listen to the changelog topic in a separate script, which is bog-standard consumer code:
```
import asyncio

import aiokafka


loop = asyncio.get_event_loop()


async def eat():
    consumer = aiokafka.AIOKafkaConsumer(
        "debug-changelog-table-changelog",
        loop=loop,
        group_id="whatever",
    )

    await consumer.start()

    try:
        async for msg in consumer:
            print(msg.value)
    finally:
        consumer.stop()


loop.run_until_complete(eat())
```

So, that's only one event per second. Still, in testing, all versions since the introduction of `send_soon()` up to your fix at 4c541cb output zero messages to the changelog, all the way until I shut down the worker, when they all flush at once.

Now 4c541cb seems to say "flush if there are 100 messages", but in fact it produces a nice and steady trickle of messages. How does this work? Thanks!

  * Python version: 3.7.2
  * Operating system: MacOS 10.14.5
  * Kafka version: 2.2.0
